general/mem_manager.hpp

```diff
@@ -13,33 +13,462 @@
+   CUDA,    ///< Memory types: { CUDA, CUDA_UVM }
+   CUDA_UVM ///< Memory types: { CUDA_UVM }
+};
+
+/// Return true is the given memory type is in MemoryClass::HOST.
+inline bool IsHostMemory(MemoryType mt) { return mt <= MemoryType::HOST_64; }
+
+/// Return a suitable MemoryType for a given MemoryClass.
+MemoryType GetSuitableMemoryType(MemoryClass mc);
+
+/// Return a suitable MemoryClass from a pair of MemoryClass%es.
+/** Note: this operation is commutative, i.e. a*b = b*a, associative, i.e.
+    (a*b)*c = a*(b*c), and has an identity element: MemoryClass::HOST.
+
+    Currently, the operation is defined as a*b := max(a,b) where the max
+    operation is based on the enumeration ordering:
+
+        HOST < HOST_32 < HOST_64 < CUDA < CUDA_UVM. */
+MemoryClass operator*(MemoryClass mc1, MemoryClass mc2);
+
+/// Class used by MFEM to store pointers to host and/or device memory.
+/** The template class paramter, T, must be a plain-old-data (POD) type.
+
+    In many respects this class behaves like a pointer:
+    * When destroyed, a Memory object does NOT automatically delete any
+      allocated memory.
+    * Only the method Delete() will deallocate a Memory object.
+    * Other methods that modify the object (e.g. New(), Wrap(), etc) will simply
+      overwrite the old contents.
+    * One difference with a pointer is that a const Memory object is NOT like a
+      const pointer: it cannot be used to overwrite entries in the memory.
+
+    A Memory object stores up to two different pointers: one host pointer (with
+    MemoryType from MemoryClass::HOST) and one device pointer (currently one of
+    MemoryType::CUDA or MemoryTyep::CUDA_UVM).
+
+    A Memory object can hold (wrap) an externally allocated pointer with any
+    given MemoryType.
+
+    Access to the content of the Memory object can be requested with any given
+    MemoryClass through the methods ReadWriteAccess(), ReadAccess(), and
+    WriteAccess(). Requesting such access may result in additional (internally
+    handled) memory allocation and/or memory copy.
+    * When ReadWriteAccess() is called, the returned pointer becomes the only
+      valid pointer.
+    * When ReadAccess() is called, the returned pointer becomes valid, however
+      the other pointer (host or device) may remain valid as well.
+    * When WriteAccess() is called, the returned pointer becomes the only valid
+      pointer, however, unlike ReadWriteAccess(), no memory copy will be
+      performed.
+
+    The host memory (pointer from MemoryClass::HOST) can be accessed through the
+    inline methods: `operator[]()`, `operator*()`, the implicit conversion
+    functions `operator T*()`, `operator const T*()`, and the explicit
+    conversion template functions `operator U*()`,  `operator const U*()` (with
+    any suitable type U). In certain cases, using these methods may have
+    undefined behavior, e.g. if the host pointer is not currently valid. */
+template <typename T>
+class Memory
```

Comment:
Somehow I think `Buffer` would be a better name, but I guess `Memory` is OK too (buffer maybe sounds too much like OpenGL).

Reply:
I guess this choice was heavily influenced by `occa::memory`. 😄